framework/infrastructure/gcp/cloud_run.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2025 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import dataclasses
+import logging
+from typing import Any, Dict, Final
+
+from googleapiclient import discovery
+
+from framework.infrastructure import gcp
+
+logger = logging.getLogger(__name__)
+
+# Type aliases
+GcpResource = gcp.compute.ComputeV1.GcpResource
+
+DEFAULT_TEST_PORT: Final[int] = 8080
+DISCOVERY_URI: Final[str] = "https://run.googleapis.com/$discovery/rest?"
+
+
+@dataclasses.dataclass(frozen=True)
+class CloudRun:
+    service_name: str
+    url: str
+
+    @classmethod
+    def from_response(cls, name: str, d: Dict[str, Any]) -> "CloudRun":
+        return cls(
+            service_name=name,
+            url=d["urls"],
+        )
+
+
+class CloudRunApiManager(
+    gcp.api.GcpStandardCloudApiResource, metaclass=abc.ABCMeta
+):
+    project: str
+    region: str
+    _parent: str
+    service: discovery.Resource
+    api_manager: gcp.api.GcpApiManager
+
+    def __init__(self, project: str, region: str):
+        if not project:
+            raise ValueError("Project ID cannot be empty or None.")
+        if not region:
+            raise ValueError("Region cannot be empty or None.")
+        self.api_manager = gcp.api.GcpApiManager(v2_discovery_uri=DISCOVERY_URI)
+        self.project = project
+        self.region = region
+        service: discovery.Resource = self.api_manager.cloudrun("v2")
+        self.service = service
+        self._parent = f"projects/{self.project}/locations/{self.region}"
+        super().__init__(self.service, project)
+
+    @property
+    def api_name(self) -> str:
+        """Returns the API name for Cloud Run."""
+        return "run"
+
+    @property
+    def api_version(self) -> str:
+        """Returns the API version for Cloud Run."""
+        return "v2"
+
+    def create_service(
+        self, service: discovery.Resource, service_name: str, body: dict
+    ) -> GcpResource:
+        return self._create_resource(
+            collection=service.projects().locations().services(),
```

Comment:
Because `self.api.projects().locations()` is used everywhere, I create a shortcut for it like so: https://github.com/grpc/psm-interop/blob/12922820b1d662b8a1b7a11d9199c4c606a49923/framework/infrastructure/gcp/network_security.py#L111

And then call `self._api_locations.servicename()`, like so: https://github.com/grpc/psm-interop/blob/12922820b1d662b8a1b7a11d9199c4c606a49923/framework/infrastructure/gcp/network_security.py#L159

Also because you only use `self.api.projects().locations().services()` you can even store it as `self._services_collection`